### DISCUSSION TITLE
Directive doesn't work with x-model

### DISCUSSION THREAD

**Discussion by guiiamorim** [alpinejs/alpine#3220] [7. Bugs (archived)] [answered]
Hi there,

I'm integrating Alpine on an existing application which already has various masks for input, so I'm trying to integrate this masks into Alpine by applying a directive on the inputs.

The masks are in vanilla javascript, I looked in the mask plugin code and maybe I'm missing something, but ultimately what it does is change the input value, so my code should work as well.

I made a pen demonstrating my problem, if I put a `x-model` on the input, my directive doesn't work, but if hook the value with a `@input` it does update the value correctly. Am I doing something wrong? I really don't want to use the mask plugin as I would have to rewrite much of my code. Any help is appreciated.

Link to my pen: https://codepen.io/guiiamorim/pen/eYrwVXe

**Comment by SimoTod** [comment, marked as answer]
Your code generates a race condition with x-model.
If you want to hook into the x-model behaviour, you need to remove the default listener and use _x_model.set using the modified value (like the x-mask plugin does). Your directive would be something like

```js
Alpine.directive(`inputmask`, (el, { value }, { effect }) => {
  const processInput = (el) => {
    effect(() => {
      if (el._x_model) {
        el._x_model.set(el.value.toUpperCase())
      } else {
        el.value = el.value.toUpperCase();
      }
    });
    if (el._x_removeModelListeners && el._x_removeModelListeners['default']) {
      el._x_removeModelListeners['default']()
      delete el._x_removeModelListeners['default']
    }
  };

  processInput(el);

  el.addEventListener("input", () => processInput(el));
});
```

**Reply by guiiamorim** [reply]
Thank you for your help, it worked like a charm!

If I may ask, how the mask plugin avoids this (does it?)? I couldn't find any code resembling what you sent me. At some point it uses the `Alpine.dontAutoEvaluateFunctions()` method, but even using it my directive didn't work.

**Reply by SimoTod** [reply]
Sorry, I said x-mask but I got confused. It's x-modelable that hooks in the model in this way. I think x-mask doesn't really work well with x-model at the moment (I might be wrong though).

**Reply by SimoTod** [reply]
Now that I'm looking at the example. You don't need effect since your code is triggered by the inpit event rather than the Alpine reactivity. You can also mode the if that remove the default listener outside the callback. Ideally you should also import the cleanup function an remove your custom listener so the directive can clean up its side effect when the element is removed from the DOM.

**Reply by guiiamorim** [reply]
Yes, you are right. It worked just fine without the effect. Thanks for the support!
